server/mqtt.go

```diff
@@ -2202,97 +2271,99 @@ func (as *mqttAccountSessionManager) processSubs(sess *mqttSession, c *client,
 		sess.cons[sid] = cc
 	}
 
-	// Helper that sets the sub's mqtt fields and possibly serialize retained messages.
-	// Assumes account manager and session lock held.
-	setupSub := func(sub *subscription, qos byte) {
-		subs := []*subscription{sub}
-		if len(sub.shadow) > 0 {
-			subs = append(subs, sub.shadow...)
+	// Helpers that sets the sub's mqtt fields and possibly serialize
+	// (pre-loaded) retained messages. Assumes account manager and session lock
+	// held.
+	initSub := func(sub *subscription, qos byte) {
+		if sub.mqtt == nil {
+			sub.mqtt = &mqttSub{}
 		}
-		for _, sub := range subs {
-			if sub.mqtt == nil {
-				sub.mqtt = &mqttSub{}
-			}
-			sub.mqtt.qos = qos
-			sub.mqtt.reserved = isMQTTReservedSubscription(string(sub.subject))
-			if fromSubProto {
-				as.serializeRetainedMsgsForSub(sess, c, sub, trace)
-			}
+		sub.mqtt.qos = qos
+		sub.mqtt.reserved = isMQTTReservedSubscription(string(sub.subject))
+	}
+
+	lockAndInitializeSub := func(subject, sid []byte, qos byte, h msgHandler) (*subscription, error) {
+		as.mu.Lock()
+		defer as.mu.Unlock()
+		sess.mu.Lock()
+		defer sess.mu.Unlock()
+
+		sub, err := c.processSub(subject, nil, sid, h, false)
+
+		if err != nil {
+			// c.processSub already called c.Errorf(), so no need here.
+			return nil, err
 		}
+		initSub(sub, qos)
```

Comment:
We could "inline" the content of initSub here by simply doing something like:
```
subs := []*subscription{sub}
subs = append(subs, sub.shadow...)
for _, sub := range subs {
    if sub.mqtt == nil {
        sub.mqtt = &mqttSub{}
    }
    sub.mqtt.qos = qos
    sub.mqtt.reserved = isMQTTReservedSubscription(string(sub.subject))
}
```

Reply:
Done in https://github.com/nats-io/nats-server/pull/4760/commits/8d6a98f4ffb7bfad8dad4243c449aa6e438db0e1